File: matlab/simplified/dpc_simple_draw.m
function dpc_simple_draw(times, X, x_min, x_max, p)
  
  x0 = X(1,:)';
  
  % this function is defined below
  [p_c, p_1, p_2] = dpc_endpositions(x0(1), x0(2), x0(3), p);
  
  figure;
  hold on;

  % draw position limits
  line([x_min(1) x_max(1)], [0 0], 'color', 'k', 'Linewidth',1.5); hold on;
  line([x_min(1) x_min(1)], [-0.1 0.1], 'color', 'k', 'Linewidth',1.5); hold on;
  line([x_max(1) x_max(1)], [-0.1 0.1], 'color', 'k', 'Linewidth',1.5); hold on;

  timer_handle = text(-0.3, x_max(1), '0.00 s','FontSize', 15);
  cart_handle = plot(p_c(1), p_c(2), 'ks', 'MarkerSize', 20, 'Linewidth', 3);
  pole_one_handle = line([p_c(1) p_1(1)], [p_c(2) p_1(2)], 'color', [38,124,185]/255, 'Linewidth', 8);
  pole_two_handle = line([p_1(1) p_2(1)], [p_1(2) p_2(2)], 'color', [38,124,185]/255, 'Linewidth', 8);
  
  joint_zero_handle = plot(p_c(1), p_c(2), 'ko', 'MarkerSize', 5);
  joint_one_handle = plot(p_1(1), p_1(2), 'ko', 'MarkerSize', 5);
  joint_two_handle = plot(p_2(1), p_2(2), 'ko', 'MarkerSize', 5);
  
  grid on;
  xlim([x_min(1)-p.r_1-p.r_2 x_max(1)+p.r_1+p.r_2]);
  ylim([x_min(1)-p.r_1-p.r_2 x_max(1)+p.r_1+p.r_2]);

  hold off;
  pause(0.5)

  for k=2:size(X,1)
    tic;
    
    x = X(k,:)';
    
    [p_c, p_1, p_2] = dpc_endpositions(x(1), x(2), x(3), p);

    set(timer_handle, 'String', sprintf('%.2f s', times(k)));

    set(cart_handle, 'Xdata', x(1))

    set(pole_one_handle, 'Xdata', [p_c(1) p_1(1)]);
    set(pole_one_handle, 'Ydata', [p_c(2) p_1(2)]);

    set(pole_two_handle, 'Xdata', [p_1(1) p_2(1)]);
    set(pole_two_handle, 'Ydata', [p_1(2) p_2(2)]); 
    
    set(joint_zero_handle, 'Xdata', p_c(1));
    set(joint_zero_handle, 'Ydata', p_c(2));
    
    set(joint_one_handle, 'Xdata', p_1(1));
    set(joint_one_handle, 'Ydata', p_1(2));
    
    set(joint_two_handle, 'Xdata', p_2(1));
    set(joint_two_handle, 'Ydata', p_2(2));
    
    pause(times(k)-times(k-1)-toc);
  end
  
end

function [p_c, p_1, p_2] = dpc_endpositions(q_0, q_1, q_2, p) 
  % Returns the positions of cart, first joint, and second joint
  % to draw the black circles
  p_c = [q_0; 0];
  p_1 = p_c + p.r_1 * [cos(q_1); sin(q_1)];
  p_2 = p_c + p.r_1 * [cos(q_1); sin(q_1)] + p.r_2 * [cos(q_1+q_2); sin(q_1+q_2)];
end

File: python/simplified/dpc_simple_draw.py
import time
import numpy as np
from matplotlib import pyplot as plt


def dpc_draw(tdata, X, x_min, x_max, p):

  x0 = X[:,1];

  # this function is defined below
  [p_c, p_1, p_2] = dpc_endpositions(x0[1], x0[2], x0[3], p)

  fig, ax = plt.subplots(1, 1)
  ax.set_aspect('equal')
  print(p)
  ax.set_xlim(x_min[1] - p["r_1"] - p["r_2"], x_max[1] + p["r_1"] + p["r_2"])
  ax.set_ylim(x_min[1] - p["r_1"] - p["r_2"], x_max[1] + p["r_1"] + p["r_2"])
  plt.show(False)
  plt.draw()
  plt.pause(0.5)

  timer_handle = plt.text(-0.3, x_max[0], '0.00 s', fontsize=15);
  cart_handle, = plt.plot(p_c[0], p_c[1], 'ks', markersize=20, linewidth=3);
  pole_one_handle, = plt.plot([p_c[0], p_1[0]], [p_c[1], p_1[1]], color=np.array([38,124,185])/255, linewidth=8);
  pole_two_handle, = plt.plot([p_1[0], p_2[0]], [p_1[1], p_2[1]], color=np.array([38,124,185])/255, linewidth=8);

  joint_zero_handle, = plt.plot(p_c[0], p_c[1], 'ko', markersize=5);
  joint_one_handle, = plt.plot(p_1[0], p_1[1], 'ko', markersize=5);
  joint_two_handle, = plt.plot(p_2[0], p_2[1], 'ko', markersize=5);

  for k in range(0, X.shape[1]):
    tic = time.time()

    x = X[:,k]

    [p_c, p_1, p_2] = dpc_endpositions(x[0], x[1], x[2], p)

    timer_handle.set_text('{:.2f} s'.format(tdata[k]))

    cart_handle.set_data(x[0], 0)

    pole_one_handle.set_data([p_c[0], p_1[0]], [p_c[1], p_1[1]])
    pole_two_handle.set_data([p_1[0], p_2[0]], [p_1[1], p_2[1]])

    joint_zero_handle.set_data(p_c[0], p_c[1]);
    joint_one_handle.set_data(p_1[0], p_1[1]);
    joint_two_handle.set_data(p_2[0], p_2[1]);

    time.sleep(np.max([tdata[k] - tdata[k-1] + time.time() - tic, 0]))
    plt.pause(0.0001)

  plt.close(fig)

def dpc_endpositions(q_0, q_1, q_2, p):
  # Returns the positions of cart, first joint, and second joint
  # to draw the black circles
  p_c = np.array([q_0, 0]);
  p_1 = p_c + p["r_1"] * np.array([np.cos(q_1), np.sin(q_1)]);
  p_2 = p_c + p["r_1"] * np.array([np.cos(q_1), np.sin(q_1)]) + p["r_2"] * np.array([np.cos(q_1+q_2), np.sin(q_1+q_2)]);
  return p_c, p_1, p_2
